File: backend/src/main/java/com/sisvendas/sisvendas/service/BaseService.java
package com.sisvendas.sisvendas.service;

import java.util.List;
import java.util.Optional;

import org.hibernate.annotations.processing.Find;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.jpa.repository.JpaRepository;

import com.sisvendas.sisvendas.models.BaseModel;

public abstract class BaseService<TEntity extends BaseModel,
    IRepository extends JpaRepository<TEntity, Long>> {
    
    @Autowired
    private IRepository repository;

    public List<TEntity> RetornarTodos() {
        List<TEntity> result = repository.findAll();
        return result;
    }

    public TEntity RetornarPorId(long id) {
        Optional<TEntity> entity = repository.findById(id);
        TEntity result = entity != null ? entity.get() : null;
        return result;
    }

    public TEntity Adicionar(TEntity entity) {
        entity.setId(0);
        repository.save(entity);
        return entity;
    }

    public TEntity Atualizar(long id, TEntity entity) {
        entity.setId(id);
        repository.save(entity);
        return entity;

    }

    public void Excluir(long id) {
        TEntity entity = RetornarPorId(id);

        if (entity != null) {
            repository.delete(entity);
        }
    }
}

File: frontend/src/services/base.service.ts
import { HttpClient } from "@angular/common/http";
import { Observable } from "rxjs";
import { BaseModel } from "../models/base.model";

export abstract class BaseService<T extends BaseModel> {
    private baseRoute = "http://localhost:8080"
    protected abstract modulo: String;

    public get route() : string {
        return `${this.baseRoute}/${this.modulo}`;
    }

    constructor(private httpClient: HttpClient) {
    }

    public retornarTodos(): Observable<T[]> {
        let result = this.httpClient.get<T[]>(this.route);
        return result;
    }

    public retornarPorId(id: Number): Observable<T> {
        let result = this.httpClient.get<T>(`${this.route}/${id}`);
        return result;
    }

    public adicionar(objeto: T) : Observable<T> {
        let result = this.httpClient.post<T>(this.route, objeto);
        return result;
    }

    public atualizar(id: Number, objeto: T): Observable<T> {
        let result = this.httpClient.put<T>(`${this.route}/${id}`, objeto);
        return result;
    }

    public excluir(id: Number): Observable<Object> {
        let result = this.httpClient.delete(`${this.route}/${id}`);
        return result;
    }
}